### DISCUSSION TITLE
I found a bug in Yalmip when the optimization variable is declared as a square matrix

### DISCUSSION THREAD

**Discussion by 871686474** [yalmip/YALMIP#1029] [General discussion]
Dear developer：
I found a bug in Yalmip~
When the optimization variable is declared as a square matrix, Yalmip will not convert the model correctly~
My code is a MPC problem, I removed all constraints (in function "solve") and give a cost function to make the first three optimization variable "x"  for "para.N" prediction time domain track the  given trajectory "para.r".  If "x"  is a square matrix，it can not find a correct solution, else, it can.  I have tried different solver and get the same result~
1. For code, you should run function "main", the gif cannot track "para.r"  correctly~
2. Then, change "x = sdpvar(para.N+1,6);" in function "solve" to "x = sdpvar(para.N+1,3);" or "x = sdpvar(para.N+1,4);" or "x = sdpvar(para.N+1,5);"...., the gif can track "para.r"  correctly~
3. You can change "para.N" in function "parain" to make "x = sdpvar(para.N+1,...);" be a square matrix with another dimension~
4. You can change the solver to have a try~
5. It is better to make a breakpoint at line 34 in function "slove" to debug~

Thanks again to the developers for their careful development of Yalmip~~~
            Student Deng Y.S.
            2021.09.
[yalmip_bug.zip](https://github.com/yalmip/YALMIP/files/7434663/yalmip_bug.zip)
28

**Comment by johanlofberg** [comment]
Without looking at the code, my guess is that you've missed a 'full' flag.
https://yalmip.github.io/tutorial/basics/

**Comment by 871686474** [comment]
Thankyou! I got it!
%&nbsp; &nbsp;You can create a sdpvar variable by:
%&nbsp; &nbsp; &nbsp;X = SDPVAR(n)&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;Symmetric nxn matrix
%&nbsp; &nbsp; &nbsp;X = SDPVAR(n,n)&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;Symmetric nxn matrix
%&nbsp; &nbsp; &nbsp;X = SDPVAR(n,m)&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;Full nxm matrix (n~=m)
%
%&nbsp; &nbsp;Definition of multiple scalars can be simplified
%&nbsp; &nbsp; &nbsp;SDPVAR x y z w
%
%&nbsp; &nbsp;The parametrizations supported are
%&nbsp; &nbsp; &nbsp;X = SDPVAR(n,n,'full')&nbsp; &nbsp; &nbsp; Full nxn matrix
%&nbsp; &nbsp; &nbsp;X = SDPVAR(n,n,'symmetric') Symmetric nxn matrix
%&nbsp; &nbsp; &nbsp;X = SDPVAR(n,n,'diagonal')&nbsp; Diagonal matrix
%&nbsp; &nbsp; &nbsp;X = SDPVAR(n,n,'toeplitz')&nbsp; Symmetric Toeplitz
%&nbsp; &nbsp; &nbsp;X = SDPVAR(n,n,'hankel')&nbsp; &nbsp; Unsymmetric Hankel (zero below the first anti-diagonal)
%&nbsp; &nbsp; &nbsp;X = SDPVAR(n,n,'rhankel')&nbsp; &nbsp;Symmetric Hankel
%&nbsp; &nbsp; &nbsp;X = SDPVAR(n,n,'skew')&nbsp; &nbsp; &nbsp; Skew-symmetric
%&nbsp; &nbsp; &nbsp;X = SDPVAR(n,n,'diagonal')&nbsp; Diagonal



------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "yalmip/YALMIP"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年10月28日(星期四) 晚上10:17
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [yalmip/YALMIP] I found a bug in Yalmip when the optimization variable is declared as a square matrix (Discussion #1029)





 
Without looking at the code, my guess is that you've missed a 'full' flag.
 https://yalmip.github.io/tutorial/basics/
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.
